crates/doc/src/heap.rs

```diff
@@ -0,0 +1,136 @@
+use super::{AsNode, BumpStr, BumpVec, Field, Fields, Node};
+
+/// HeapDoc is a document representation stored in the heap.
+#[derive(Debug, rkyv::Archive, rkyv::Serialize)]
+#[archive(archived = "ArchivedDoc")]
+pub struct HeapDoc<'alloc> {
+    /// Root node of the document.
+    pub root: HeapNode<'alloc>,
+    /// Arbitrary flags used to persist document processing status.
+    pub flags: u8,
+}
+
+/// HeapNode is a document node representation stored in the heap.
+// The additional archive bounds are required to satisfy the compiler due to
+// the recursive nature of this structure. For more explanation see:
+// https://github.com/rkyv/rkyv/blob/master/examples/json/src/main.rs
+#[derive(Debug, rkyv::Archive, rkyv::Serialize)]
+#[archive(
+    archived = "ArchivedNode",
+    bound(
+        serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer + rkyv::ser::SharedSerializeRegistry"
+    )
+)]
+pub enum HeapNode<'alloc> {
+    Array(#[omit_bounds] BumpVec<'alloc, HeapNode<'alloc>>),
+    Bool(bool),
+    Bytes(BumpVec<'alloc, u8>),
+    Float(f64),
+    NegInt(i64),
```

Comment:
I take it the reason for separate +/- int representations is to allow representing unsigned integers that are larger than the `i64` max?

Reply:
Correct. The precedent here is our `json::Number` type which is itself modeled on `serde_json::Value`.

`HeapNode` flattens it in this way 'cause it turns out that saves a bunch of space in the archived representation, as `rkyv` isn't smart enough to smash together multiple nested enum discriminants